### PR TITLE
Support for frontchannel_logout_session_required OIDC client parameter

### LINKED ISSUE #10137: Support for frontchannel_logout_session_required OIDC client parameter
### Description

Keycloak currently does not have support for this parameter, which is mentioned in the OIDC specification https://openid.net/specs/openid-connect-frontchannel-1_0.html#OPLogout

This JIRA involves implement support for this parameter to both OIDC dynamic client registration and Keycloak admin console.

Also OIDC Well-Known endpoint needs to be updated to indicate support for frontchannel logout.

This is blocker for OIDC certification with frontchannel logout profile (See issues 5 and 6 in https://docs.google.com/document/d/1yKg_F-pG0NPz3pxLtDc7g288d7jAC8f0leABh50BZn0/edit# )

### Discussion

_No response_

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by mposolda**
Closes #10137

### CODE REVIEW

**Comment by mposolda** [comment]
@tnorimat @stianst Do you have a chance to review please? This parameter is needed for the OIDC frontchannel logout certification

File: services/src/main/java/org/keycloak/protocol/oidc/OIDCAdvancedConfigWrapper.java [hunk 1]
```diff
@@ -321,6 +321,17 @@ public String getFrontChannelLogoutUrl() {
         return getAttribute(OIDCConfigAttributes.FRONT_CHANNEL_LOGOUT_URI);
     }
 
+    public boolean isFrontChannelLogoutSessionRequired() {
+        String frontChannelLogoutSessionRequired = getAttribute(OIDCConfigAttributes.FRONT_CHANNEL_LOGOUT_SESSION_REQUIRED);
+        // Include session by default for backwards compatibility
+        return frontChannelLogoutSessionRequired == null ? true : Boolean.parseBoolean(frontChannelLogoutSessionRequired);
```
Comment (tnorimat) [inline services/src/main/java/org/keycloak/protocol/oidc/OIDCAdvancedConfigWrapper.java:327]: Accordin to https://openid.net/specs/openid-connect-frontchannel-1_0.html#RPLogout ,
It should be false if frontChannelLogoutSessionRequired is not defined.
WDYT?

File: services/src/main/java/org/keycloak/services/clientregistration/oidc/DescriptionConverter.java [hunk 2]
```diff
@@ -234,6 +234,11 @@ public static ClientRepresentation toInternal(KeycloakSession session, OIDCClien
         }
 
         configWrapper.setFrontChannelLogoutUrl(Optional.ofNullable(clientOIDC.getFrontChannelLogoutUri()).orElse(null));
+        if (clientOIDC.getFrontchannelLogoutSessionRequired() == null) {
+            configWrapper.setFrontChannelLogoutSessionRequired(true);
```
Comment (tnorimat) [inline services/src/main/java/org/keycloak/services/clientregistration/oidc/DescriptionConverter.java:238]: Accordin to https://openid.net/specs/openid-connect-frontchannel-1_0.html#RPLogout ,
It should be false if frontChannelLogoutSessionRequired is not defined.
WDYT?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/RPInitiatedLogoutTest.java [hunk 3]
```diff
@@ -627,6 +627,36 @@ public void testFrontChannelLogoutWithPostLogoutRedirectUri() throws Exception {
         }
     }
 
+    @Test
+    public void testFrontChannelLogoutWithoutSessionRequired() throws Exception {
```
Comment (tnorimat) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/RPInitiatedLogoutTest.java:631]: It seems to be better to add a test where frontchannel_logout_session_required is false.

**Review by tnorimat — COMMENTED** [review]
@mposolda I've added review comments. Could you check them?

File: services/src/main/java/org/keycloak/protocol/oidc/OIDCAdvancedConfigWrapper.java (hunk 1 above)
Comment (mposolda) [reply services/src/main/java/org/keycloak/protocol/oidc/OIDCAdvancedConfigWrapper.java:327]: Yes, however another point is backwards compatibility. The previous version of Keycloak already has support for "FrontChannelLogout", but in fact it uses "frontChannelLogoutSessionRequired" set to true as it included "sid" and "iss" claims. Previous versions of Keycloak doesn't have this attribute defined, so switching this to false by default will in fact be broken backwards compatibility behaviour for them. Unless we add hook to the migration, which itself is typically a painful and can have other side-effects on performance etc...

So I suggest to keep this to true for backwards compatibility, but use default value "false" for the case when OIDC client is created or updated through OIDC dynamic client registration. In this case, it is fine to use "false" by default as we won't break backwards compatibility. And at the same time, we will be fine with the OIDC specification and use "false" for OIDC registered client. WDYT?

File: services/src/main/java/org/keycloak/services/clientregistration/oidc/DescriptionConverter.java (hunk 2 above)
Comment (mposolda) [reply services/src/main/java/org/keycloak/services/clientregistration/oidc/DescriptionConverter.java]: Yes, agree. I changed this as described in the comments above.

I've added another commit when I changed the default to false for the OIDC dynamic client registration, so we are fine with the specification. Could you please check it? I used another separate commit just for the easier review, but I plan to squash this before merging it.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/RPInitiatedLogoutTest.java (hunk 3 above)
Comment (mposolda) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/RPInitiatedLogoutTest.java:631]: I am not sure I understand this comment?

- The `testFrontChannelLogoutWithPostLogoutRedirectUri` tests the case when  frontchannel_logout_session_required is true
- The `testFrontChannelLogoutWithoutSessionRequired` tests the case when  frontchannel_logout_session_required is false

So both cases are covered.

Or maybe I misunderstood what exactly you meant by this comment?

**Comment by mposolda** [comment]
@tnorimat Thanks for the review. I've replied to your comments and added one additional commit to this PR as described above.

File: services/src/main/java/org/keycloak/protocol/oidc/OIDCAdvancedConfigWrapper.java (hunk 1 above)
Comment (tnorimat) [reply services/src/main/java/org/keycloak/protocol/oidc/OIDCAdvancedConfigWrapper.java:327]: Yes, I agree with your proposal.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/RPInitiatedLogoutTest.java (hunk 3 above)
Comment (tnorimat) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/RPInitiatedLogoutTest.java:631]: Sorry, I've misunderstood. Your original codes are fine.

File: services/src/main/java/org/keycloak/services/clientregistration/oidc/DescriptionConverter.java (hunk 2 above)
Comment (tnorimat) [reply services/src/main/java/org/keycloak/services/clientregistration/oidc/DescriptionConverter.java]: I've checked your added commit. It is fine.

**Review by tnorimat — APPROVED** [review]
LGTM

**Comment by mposolda** [comment]
@tnorimat Thanks!

**Pull request merged**